src/distance.cpp

```diff
@@ -0,0 +1,238 @@
+#include "distance.hpp"
+using namespace std;
+namespace vg {
+
+    //////////  Helper Functions
+
+    size_t snarlMatrixIndex (unordered_map<pair<id_t, bool>, size_t>* indices, 
+                            pair<id_t, bool> start, pair<id_t, bool> end) {
+        /*Get the index of dist from start to end in a snarl distance matrix
+          given the node ids + direction */
+        unordered_map<pair<id_t, bool>, size_t> map = *indices;
+        size_t i1 = map[start];
+        size_t i2 = map[end];
+        size_t length = map.size();
+        //return (length + 1) * (length / 2) + (i1+1)* (i1/2) + i2 - 1;
+        /*TODO Currently store entire distance matrix since it's simpler
+           dist from n1 -> n2 = -n2 -> -n1 - len(n1) + len(n2) */
+        return i1 * length + i2;
+    };
+   
+    int64_t snarlDistance(SnarlDistances* sd,pair<id_t, bool> n1, 
+                                   pair<id_t, bool> n2){
+        //Distance between two nodes n1 and n2 in snarl
+        //nodes are node id- negative if reverse, positive if forward
+        //TODO: the distance index is passed as an argument for now
+        size_t i = snarlMatrixIndex(sd->visit_to_index, n1, n2);
+        vector<int64_t> distances = *sd->distances;
+        return distances[i]; 
+    };
+   
+    int64_t chainDistance(ChainDistances* cd, int v1, int v2) {
+        //Distance between v1 and v2 in a chain
+        unordered_map<id_t, size_t> snarl_to_index = *cd->snarl_to_index;
+        size_t i1 = snarl_to_index[v1];
+        size_t i2 = snarl_to_index[v2]; 
+        vector<int64_t> prefix_sum = *cd->prefix_sum;
+        return abs(prefix_sum[i1] - prefix_sum[i2]);
+    };
+
+    int64_t chainLength(DistanceIndex* di, Chain* chain) {
+        //Get the length of a chain including length of last node
+        id_t c = get_start_of(*chain).snarl().start().node_id();
+        unordered_map<id_t, ChainDistances*> cd_map = di->cd;
+        vector<int64_t>* prefix_sum = cd_map[c]->prefix_sum;
+        vector<int64_t> prefix = *prefix_sum;
+        return prefix.back();
+    };
+
+
+
+
+
+    /////////   Create the distance index
+
+
+    bool cmp (pair<const handle_t*, int64_t> x, pair<const handle_t*, int64_t> 
+           y) {
+        /*Comparison function for the priority of a pair of handle, distance*/
+        return (x.second < y.second); 
+    };
+
+    int64_t calculateIndex(DistanceIndex* di, VG* graph, SnarlManager& sm, 
+                 const Chain* chainp) {
+        Chain chain = *chainp; 
+        vector<int64_t> chain_prefix_sum (1,0); //initialize to [0]
+        unordered_map<id_t, size_t> chain_snarl_to_index;
+        for (int i = 0; i < chain.size(); i++) { 
+            const Snarl snarl = *chain[i];
```

Comment:
Here too, better to leave this a pointer. Initializing a variable with the `=` operator makes a fully copy. In the case of a pointer, that's not such a big deal because it's 64 bits of data. For larger classes, it can end up being a lot more.